### PR TITLE
[3.10] A well-known URL for changing passwords

### PR DESCRIPTION
**Pull request by zero-24**
### Summary of Changes

This here implements a well-known URL for changing passwords. https://web.dev/change-password-url/ using a htaccess rule.

### Testing Instructions

- apply the htaccess rule
- try to access example.org/.well-known/change-password/

### Actual result BEFORE applying this Pull Request

404

### Expected result AFTER applying this Pull Request

htaccess redirect to index.php?option=com_users&view=reset

image

### Documentation Changes Required

None

### web.config status

That here should work for web.config
```
               <rule name="A well-known URL for changing passwords">
                   <match url="(.*)" ignoreCase="false" />
                   <conditions logicalGrouping="MatchAll">
                     <add input="{URL}" pattern="\.well-known/change-password$" ignoreCase="true" negate="true" />
                   </conditions>
                   <action type="Redirect" url="index.php?option=com_users&view=reset" statusCode="302" />
               </rule>
```

But for some reason this validator here: https://elmah.io/tools/configvalidator/ seems to not like the new URL.
image

Any idea how to fix or test it? Else I would suggest to add a note to the postinstall that for web.config there is no known workaround yet.

### CODE REVIEW

**Comment by HLeithner** [comment]
I expect that this only works for root installations and not in subfolders? if so we should mention this.

**Comment by SharkyKZ** [comment]
>Any idea how to fix or test it?

Escape ampersand.

**Comment by zero-24** [comment]
> I expect that this only works for root installations and not in subfolders? if so we should mention this.

The redirect works on both.

**Comment by zero-24** [comment]
> >Any idea how to fix or test it?
> 
> Escape ampersand.

The validator is not happy with the `=` or i'm missing somsthing?

**Comment by SharkyKZ** [comment]
No, it's the ampersand. You're editing a XML file.

**Comment by zero-24** [comment]
Hmm that did the trick aktually.. i could swear that i have tried that before too..

**Comment by zero-24** [comment]
Pushed the fix :-)

**Comment by SharkyKZ** [comment]
> 
> 
> > I expect that this only works for root installations and not in subfolders? if so we should mention this.
> 
> The redirect works on both.

The redirect works but this feature will not be detected because `.well-known` has to be in the root directory.

**Comment by zero-24** [comment]
Is there anything we can do about that?

Or do you have a suggestion for an text to be added?

**Comment by brianteeman** [comment]
I am not in favour of this. It should be up to the site owner if users are able to reset their passwords. There are lots of business cases where it is not allowed.

**Comment by zero-24** [comment]
Ok fine for me than we can set it up commented. Do you have a suggestion for the text?

**Comment by brianteeman** [comment]
also see JSST report

**Comment by HLeithner** [comment]
it's to the side owner if s/he adds the htaccess redirect or not. So basically it's a good thing so people use password mangers and have a easy way to finde the change form.

and it doesn't change anything to the functionality because you always can access change or reset functions if not disabled in a proper way.

**Comment by brianteeman** [comment]
>  it's to the side owner if s/he adds the htaccess redirect or not

On updates yes but on new installs it would be the default, and unexpected, behaviour that really should not be in a patch release

**Comment by wilsonge** [comment]
Is there a reason to put this in .htaccess anyhow and not a plugin which manipulates the Router? Would be a good example of a router plugin and make it more easy to toggle on/off

**Comment by HLeithner** [comment]
> On updates yes but on new installs it would be the default, and unexpected, behaviour that really should not be in a patch release

I don't have your opinion on this because it basically does nothing new, except giving a hint for software that is/should already known to Joomla users. The SVG htaccess addition is much more problematic then this redirect.

> Is there a reason to put this in .htaccess anyhow and not a plugin which manipulates the Router? Would be a good example of a router plugin and make it more easy to toggle on/off

Hard to say, basically you can do it with com_redirect and don't need any plugin or htaccess redirect but tbh post site creates didn't know that this .well-known path exists (not even I had known this) and I use many well known pathes ;-)

**Comment by zero-24** [comment]
> Is there a reason to put this in .htaccess anyhow and not a plugin which manipulates the Router? Would be a good example of a router plugin and make it more easy to toggle on/off

Hmm i thourgth about a plugin or com_redirect for redirect the problem is thats disabled by default. I'm fine with an dedicated plugin but is't that an overkill for such a minimal redirect?

**Comment by zero-24** [comment]
> > On updates yes but on new installs it would be the default, and unexpected, behaviour that really should not be in a patch release
> 
> I don't have your opinion on this because it basically does nothing new, except giving a hint for software that is/should already known to Joomla users. The SVG htaccess addition is much more problematic then this redirect.

Agree it is in the end just an alias for com_users&view=reset. I'm also fine to move that into 3.10 when there are concerns that it would be an issue in an Patch release.

**Comment by gostn** [comment]
makes it sense to test?

**Comment by HLeithner** [comment]
.htaccess is the more straight forward way, adding a plugin is an overkill (as example good of course), a default com_redirect rule doesn't work for people who doesn't use it (like me^^).

We can uncomment it per default  but I don't see a reason to hide it because this url is "well-known" to people who think it's an attack vector...

Can you rebase it on 3.10 and @gostn can test it with 3.10 (or now because it's version independent)

**Comment by zero-24** [comment]
> Can you rebase it on 3.10 and @gostn can test it with 3.10 (or now because it's version independent)

I dont know why this should wait for 3.10 but fine for me, done.

**Comment by zero-24** [comment]
Ah have to change the sql files too .. give me a few minutes ..

**Comment by zero-24** [comment]
Files has been moved to its dedicated 3.10.0 file so this is ready for testing.

**Comment by HLeithner** [comment]
> > Can you rebase it on 3.10 and @gostn can test it with 3.10 (or now because it's version independent)
> 
> I dont know why this should wait for 3.10 but fine for me, done.

mainly to not force a manual not really necessary step on every bugfix release.

**Comment by brianteeman** [comment]
and semantic versioning rules say no new features in a bug fix release

File: administrator/language/en-GB/en-GB.com_admin.ini [hunk 1]
```diff
@@ -138,6 +138,8 @@ COM_ADMIN_PHPINFO_DISABLED="The built in phpinfo() function has been disabled by
...
 COM_ADMIN_POSTINSTALL_MSG_HTACCESS_AUTOINDEX_TITLE=".htaccess Update Concerning Directory Listings"
+COM_ADMIN_POSTINSTALL_MSG_HTACCESS_CHANGEPWURL_DESCRIPTION="<p>Starting with 3.9.22 the default htaccess.txt file contains a specific rule to support <a href='https://web.dev/change-password-url/' target='_blank' rel='noopener noreferrer'>a well-known URL for changing passwords</a>. We recommend to manually apply the necessary changes to any existing .htaccess file, as this file can not be updated automatically.</p><p>The new lines are:</p><pre>## Help users change passwords easily by adding a well-known URL for changing passwords\n## https://web.dev/change-password-url/\nRewriteCond %{REQUEST_URI} \.well-known/change-password$\nRewriteRule .* index.php?option=com_users&view=reset [R=302,L]</pre>\nThat lines should be added just before this line in the existing files:<pre>## Begin - Custom redirects</pre>"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_admin.ini]:
```suggestion
COM_ADMIN_POSTINSTALL_MSG_HTACCESS_CHANGEPWURL_DESCRIPTION="<p>Starting with 3.10 the default htaccess.txt file contains a specific rule to support <a href='https://web.dev/change-password-url/' target='_blank' rel='noopener noreferrer'>a well-known URL for changing passwords</a>. We recommend to manually apply the necessary changes to any existing .htaccess file, as this file can not be updated automatically.</p><p>The new lines are:</p><pre>## Help users change passwords easily by adding a well-known URL for changing passwords\n## https://web.dev/change-password-url/\nRewriteCond %{REQUEST_URI} \.well-known/change-password$\nRewriteRule .* index.php?option=com_users&view=reset [R=302,L]</pre>\nThat lines should be added just before this line in the existing files:<pre>## Begin - Custom redirects</pre>"
```

**Comment by gostn** [comment]
I have tested this item :red_circle: unsuccessfully on b1263088db907604011db786a28a0525cc554694<br /><br />´404 Category not found´ using PR.

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30750">issues.joomla.org/tracker/joomla-cms/30750</a>.</sub>

**Comment by zero-24** [comment]
@gostn has the .htaccess file enabled and extended with the rule in this PR? What URL did you choose?

**Comment by gostn** [comment]
> .htaccess file enabled and extended with the rule in this PR?

yes, followed the instructions.

> What URL did you choose?

a test-installation on cloudaccess (deleted now).

**Pull request closed without merging**